### PR TITLE
Enable multiple grid reductions when possible

### PR DESCRIPTION
**Pull request by csarofeen**
Allows multiple grid reductions if there is not a dependency between them. Primarily the argument ordering in executor had to be fixed since the codegen was running through allocations as they appeared, and executor was pushing all zero init buffers to the end. Ordering in executor is now the same as codegen.

For now I'm not hooking this up to multi-reduction scheduler as the one case I was hoping to see big improvements doesn't show any. For now leaving it enabled so we have this tool for performance.

### CODE REVIEW

File: test/cpp/jit/test_gpu.cpp [hunk 1]
```diff
@@ -13078,15 +13113,14 @@ TEST(NVFuserTest, FusionMultipleGridReductions_CUDA) {
   Fusion fusion;
```
Comment (shmsong) [inline test/cpp/jit/test_gpu.cpp]: nitpick: comment above this one doesn't seem related. Also does this test cover a scenario that `FusionMultiGridReduction` doesn't?

File: torch/csrc/jit/codegen/cuda/codegen.cpp [hunk 2]
```diff
@@ -836,8 +836,8 @@ class CudaKernelGenerator : private kir::IrVisitor {
              << "reduction::gridReduce<" << flags_str << ">(\n";
     indent() << kTab << gen(rop->out()) << ",\n";
     if (domain->hasBlockReduction()) {
-      indent() << kTab << "block_result"
-               << ",\n";
+      indent() << kTab << "block_result_" << block_reduce_name_ << ",\n";
+      block_reduce_name_++;
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/codegen.cpp:840]: nitpick/minor question: why do we want to increment the `block_reduce_name_` in `GridReduction` instead of doing it in code-gen of the block reduction itself? This will assume a particular ordering between the multiple block and grid reductions.

Related to this, maybe just wondering if we'd ever want to change the ordering of block and grid reductions given that we now can have many grid reductions.

**Review by shmsong — APPROVED** [review]
Looks good to me so far. Had some minor questions. Thanks for extending the codegen and executor for multiple grid reduction support 👍 . 

Also I see this is WIP, just wondering is the `no dependency between grid reductions` part going to be in this PR or in a follow up?

File: torch/csrc/jit/codegen/cuda/codegen.cpp (hunk 2 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/codegen.cpp:840]: From what I could tell the ordering when we have a block reduction followed by a grid reduction, is block reduction will be called first with the intermediate result going into "block_result...", then grid reduction will be called consuming that "block_result..." I couldn't find any instance where we had a block+grid reduction where we didn't first call the block reduction code generator then the grid reduction code generator. So I figured it was safest to increment in the grid reduction. It also seemed to me we would never generate "block_result..." without having a block -> grid reduction. So this seems safe to me.

File: test/cpp/jit/test_gpu.cpp (hunk 1 above)
Comment (csarofeen) [reply test/cpp/jit/test_gpu.cpp]: I had a slight mistake in this test update, and forgot to update the comment. This is testing the dependency structure you mentioned above. It's making sure this scenario throws.

**Comment by csarofeen** [comment]
> Also I see this is WIP, just wondering is the `no dependency between grid reductions` part going to be in this PR or in a follow up?

I couldn't find any cases where having a dependency between grid reductions would actually compile. So it seems these cases are all covered by existing errors. When we get to cooperative groups and grid broadcasts this will have to be re-evaluated.

File: torch/csrc/jit/codegen/cuda/codegen.cpp (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/codegen.cpp:840]: This should be safe, although I think ideally the dependency through the intermediate flag should be represented in KIR.  For now, the current method seems just fine.

File: test/cpp/jit/test_gpu.cpp [hunk 3]
```diff
@@ -4579,6 +4579,58 @@ TEST(NVFuserTest, FusionReduction6_CUDA) {
...
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor input = at::randn({numel_x, numel_y}, options);
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion);
+  auto cg_outputs = fe.runFusion({input});
+
+  std::vector<at::Tensor> aten_outputs = {
+      std::get<0>(input.to(at::kDouble).max({0})), input.to(at::kDouble).sum({0})};
```
Comment (naoyam) [inline test/cpp/jit/test_gpu.cpp]: nit: `.max(0)` instead of `.max({0})`. Apparently, there's no override of `max` with `IntArrayRef`. Clang complains.

**Review by naoyam — APPROVED** [review]
LGTM
